crypto/src/traits.rs

```diff
@@ -122,14 +135,102 @@ pub trait Authenticator:
 {
     type PubKey: VerifyingKey<Sig = Self>;
     type PrivKey: SigningKey<Sig = Self>;
+    type AggregateSig: AggregateAuthenticator<Sig = Self>;
+    const LENGTH: usize;
 }
 
 /// Trait impl'd by a public / private key pair in asymmetric cryptography.
 ///
-pub trait KeyPair {
+pub trait KeyPair: Sized + From<Self::PrivKey> {
     type PubKey: VerifyingKey<PrivKey = Self::PrivKey>;
     type PrivKey: SigningKey<PubKey = Self::PubKey>;
+    type Sig: Authenticator<PubKey = Self::PubKey>;
+
     fn public(&'_ self) -> &'_ Self::PubKey;
     fn private(self) -> Self::PrivKey;
+
+    #[cfg(feature = "copy_key")]
+    fn copy(&self) -> Self;
+
     fn generate<R: CryptoRng + RngCore>(rng: &mut R) -> Self;
+    fn public_key_bytes(&self) -> <Self::PubKey as VerifyingKey>::Bytes;
+
+    fn hkdf_generate(
+        // The Input Key Message defined by the HKDF specification. 32 bytes in length.
+        ikm: &[u8],
+        // A salt for key generation
+        salt: &[u8],
+        // An optional 'info' field as defined by the HKDF specification
+        info: Option<&[u8]>,
+    ) -> Result<Self, signature::Error> {
+        if ikm.len() != 32 {
+            return Err(signature::Error::new());
+        }
+        Self::hkdf_generate_from_ikm(ikm, salt, info.unwrap_or(&DEFAULT_DOMAIN))
+    }
+
+    fn hkdf_generate_from_ikm(
+        ikm: &[u8],  // IKM (32 bytes)
+        salt: &[u8], // Optional salt
+        info: &[u8], // Optional domain
+    ) -> Result<Self, signature::Error> {
+        let hk = hkdf::Hkdf::<Sha3_256>::new(Some(salt), ikm);
+        const SHA3_OUTPUT_SIZE: usize = <Sha3_256 as OutputSizeUser>::OutputSize::USIZE;
+        let mut okm = [0u8; SHA3_OUTPUT_SIZE];
+        hk.expand(info, &mut okm)
+            .map_err(|_| signature::Error::new())?;
+
+        let secret_key = Self::PrivKey::from_bytes(ikm).map_err(|_| signature::Error::new())?;
+
+        let keypair = Self::from(secret_key);
+        Ok(keypair)
+    }
```

Comment:
I have solved the problem that this `Sha3` based implementation may not produce hashes of the suitable length for the byte representation of the private key, in https://github.com/punwai/narwhal-1/pull/5